modules/dnn/src/cuda/types.hpp

```diff
@@ -0,0 +1,28 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef OPENCV_DNN_SRC_CUDA_TYPE_HPP
+#define OPENCV_DNN_SRC_CUDA_TYPE_HPP
+
+#include <cstdint>
+
+namespace cv { namespace dnn { namespace cuda4dnn { namespace csl {
+
+    /* For indices, we can use 32bit variables or 64bit variables. The GPU registers are 32 bits in size.
+     * Hence, a 64bit variable requires two registers and is slightly slower than the 32bit versions for some
+     * operations.
+     */
+    namespace gpu {
+#ifdef __CUDACC__
+    using size_type = int;
+    using index_type = int;
+#else
+    using size_type = std::int32_t;
```

Comment:
These aliases are used to determinate the index and size variable type that are used in the kernels. Using 32-bit variables can considerably improve the performance (epsecially in kernels with high register usage). However, this prevents the use of large tensors. If large tensor are to be supported, these aliases must be changed to a 64bit type. This will come with a performance penalty but would support large tensors.

I think it's useful to have a build option — something like `DNN_CUDA_INDEX_TYPE_64` or `DNN_CUDA_REQUIRE_LARGE_TENSOR` — to select the index sizes. This way a user who wants to run a model with huge tensors would have just set a build option instead of having to edit the code.

Currently, cuDNN doesn't allow more than 2 billion **elements** (not bytes) per tensor but the raw CUDA kernels can work with arbitrarily sized tensors as long as the indices fit in the variables. However, using 32-bit indices could potentially lead to intermediate overflow even if the element count is less than 2 billion.

Moreover, it could be possible that cuDNN might support large tensors in the future.

TODO: check kernels for intermediate overflow and try avoid wherever possible

Reply:
I generally think this is a good idea and you should do this.  However, since cudnn doesn't support 64bit indexes yet what would `DNN_CUDA_REQUIRE_LARGE_TENSOR` do?  Certainly once cudnn supports this such an option should be added, assuming there is still a performance hit for switching to 64bit vs 32bit. But can such an option do something substantive at the moment?